Control Stretchly from commandline

papey opened the pull request:
<!--

Have you read Code of Conduct? By filing an Pull Request, you are expected to comply with it, including treating everyone with respect: https://github.com/hovancik/stretchly/blob/master/CODE_OF_CONDUCT.md

-->

Issue: closes #458 
<!-- Link to relevant issue. All PRs should be associated with an issue -->

### Requirements

<!-- If there’s a checkbox you can’t complete for any reason, that's okay, just explain in detail why you weren’t able to do so. -->

- [X]  issue was opened to discuss proposed changes before starting implementation. It is important do discuss changes before implementing them (Why should we add it? How should it work? How should it look? Where will it be? ...).
- [x]  during development, `node` version specified in `package.json` was used (ie using [nvm](https://github.com/creationix/nvm)).
- [x]  package versions and package-lock.json were not changed (`npm install --no-save`).
- [x]  app version number was not changed.
- [x]  all new code has tests to ensure against regressions.
- [x] `npm run lint` reports no offenses.
- [x] `npm run test` is error-free.
- [x]  README and CHANGELOG were updated accordingly.
- [ ]  after PR is approved, all commits in it are [squashed](https://gitbetter.substack.com/p/how-to-squash-git-commits)

### Description of the Change

Contribute and finish the work started on PR #567.

### Verification Process

I've checked that commands goes from commandline call to main Stretchly instance

### Other information

none

papey commented:
Looking at the code, I see mainly two issues we need to discuss in order to get things done.

First things first, for the implementation part, after playing with multiple kind of inputs, I think this approach : https://github.com/hovancik/stretchly/pull/567#issuecomment-634654738 is the best one. It's simple and efficient, without the need to type too much extra chars.

The other one is about how commands should be handled.

For now, we are only passing args to the main instance using : 

```app.on('second-instance', runCommand)``` 

Using the `second-instance` event, `stdout` is bind to the main instance so you can't handle commands that needs to return outputs to the shell launching the command. So what about the `help` and version `command` ?

One simple idea is to break command handling into two parts, the ones that needs to be forwarded to the main instance and the one that only outputs text to the user. With this solution, the ugly thing will be : 

* the need to parse args twice, first time in the `comandline instance` and another time in the `main instance`, in the callback
* some commands will go into the `second-instance` callback but will do nothing (`help` and `version` at least)

Thank you in advance for your highlights and guidance on this subjects.

hovancik commented:
:+1: for the interface you chose. I think when testing, you'll see if it feels natural or not. 

Regarding help and version: when I was playing with code first time around, I came to the same conclusion as you, that there should have to be 2 parsers. I haven't found other way that time.

papey commented:
Consider this as WIP, i'm trying to find the best way to handle this but a lib might be a good option.

papey commented:
Here is a the help message for this first draft

```
Stretchly is already running, sending command

Usage: stretchly <command> [options]

Commmands:
        stretchly help            Show this help message
        stretchly version         Show current stretchly version
        stretchly reset           Reset breaks
        stretchly pause [options] Pause breaks
        stretchly resume          Resume from a pause
        stretchly toggle          Toggle breaks between resume/paused
        stretchly mini [options]  Skips to next Mini Break
        stretchly long [options]  Skips to next Long Break

Options:
        -T, --title    Specify title for next break (Long or Mini)
        -t, --text     Specify text for next break (Long Break only)
        -s, --skip     Skip directly to this break (Long or Mini)
        -d, --duration Specify duration for pausing breaks (Pause only)
```

For now, everything is done without any parsing lib. I've made some assumptions in order to keep things simple in the command handling code since we need to handle commands twice (one in the `second instance` one in the `main instance`).

All the important stuff is in a dedicated module `commands.js`. Even if it's quite simple, I've tried to organize all the things logically.

The most complex code is in the help message. This is a generated message from all the commands and associated options defined in the module. Handling the padding to get a nice output is what makes things a bit ugly in this code. It needs a bit of rework but it does the job for now.

I also add a `toggle` command since this is the one I need as keyboard shortcut.

hovancik commented on app/main.js:
```diff
@@ -41,7 +43,10 @@ global.shared = {
 
 if (!gotTheLock) {
-  console.log('Stretchly is already running.')
+  console.log('Stretchly is already running, sending command\n')
```
I've recently added `electron-log` package for logging. I think we could try using it for logging this as well. 

I've been using this way of logging: https://github.com/hovancik/stretchly#logging Maybe use something like `Stretchly 2nd instance: message` or sth like that. I guess it should go to same file.

papey replied:
See https://github.com/papey/stretchly/commit/554a61293032a69859973f50b6e5c52be373b4b2

hovancik replied:
Awesome :) If you see some place that you think could be good one for some debug message, feel free to add it :)

hovancik commented on app/utils/commands.js:
```diff
@@ -0,0 +1,241 @@
+    this.command = input[0]
+
+    if (!this.supported[this.command]) {
+      throw new Error(`Error: command ${this.command} is not supported`)
```
I usually run with `npm run dev`. Could we make it run fine with this? 

```
"dev": "cross-env NODE_ENV=development electron ./app --enable-logging --remote-debugging-port=9222",
```

It's throwing `Error: Error: command --enable-logging is not supported` now.

papey replied:
It's thrown if you have another primary stretchly instance running. The code pass in https://github.com/papey/stretchly/blob/features/cmdline-automation/app/main.js#L45 this path. If you need multiple stretchly instance, this `gotTheLock` might not be a good option.

hovancik replied:
What I mean is: 

This runs fine:`npm start -- pause`
This does not run fine `npm run dev -- pause`

I guess we need to add `--enable-logging --remote-debugging-port=9222` to valid comands, but ignore them.

hovancik commented on app/utils/commands.js:
```diff
@@ -0,0 +1,241 @@
+      })
+
+      if (!valid) {
+        throw new Error(`Error: options ${name} is not valid for command ${this.command}`)
```
Do we need to throw here? I don't think we should get window from command line 

![image](https://user-images.githubusercontent.com/4745859/95010888-2771da80-062d-11eb-8510-2c9b822629e7.png)

papey replied:
It's something I wanted to discuss with you, I left this on purpose in order to get a clear error. Since it's in the second instance, there is no way to get clear feedback. I can log it to stdout in the main instance. Another solution would be a notification, maybe ?

papey replied:
I replaced it with a log, see https://github.com/papey/stretchly/commit/554a61293032a69859973f50b6e5c52be373b4b2

hovancik replied:
I guess we should try to do as much of validation as possible in command instance, with errors to output. So that primary instance doesn't need to care. 

I am not sure for example what `checkInForward` is for in main.

papey replied:
I renamed it and add some comments, it's the function run in the main instance to return early if the command is a `second-instance` kind (`version` and `help` atm).

papey replied:
I fixed it naively here : https://github.com/hovancik/stretchly/pull/700/commits/42c3b2e440645432d68ea710d5d07abf38dadb4e

hovancik commented on app/utils/commands.js:
```diff
@@ -0,0 +1,253 @@
+  parse (input) {
+    // filter out electron flags first
+    let i = 0
+    while (input[i].startsWith('--')) {
```
`npm run dev` gives error

```
TypeError: Cannot read property 'startsWith' of undefined
```

papey replied:
Ok, it triggers when you have an already running instance, fixing it

papey replied:
Fixed by https://github.com/hovancik/stretchly/pull/700/commits/1424058ee0283fa2dbab94784456718c93f8908a

hovancik commented on app/utils/commands.js:
```diff
@@ -0,0 +1,253 @@
+  }
+
+  runOrForward () {
+    if (!this.command) {
```
I think we don't need this as we are making `help` the default command?

papey replied:
Fixed by https://github.com/hovancik/stretchly/pull/700/commits/c82e0d097aa1ab4d4a8c7ba6d60c2d07d7d7c223

hovancik commented on app/utils/commands.js:
```diff
@@ -0,0 +1,253 @@
+      return
+    }
+
+    if (!this.supported[this.command]) {
```
I think this will never happen as we catch this in `parse`. Let's use error there instead of throwing?

papey replied:
Fixed by https://github.com/hovancik/stretchly/pull/700/commits/c82e0d097aa1ab4d4a8c7ba6d60c2d07d7d7c223

hovancik commented on app/main.js:
```diff
@@ -915,6 +923,47 @@ function showNotification (text) {
+
+    case 'mini':
+      if (cmd.options.title) nextIdea = [cmd.options.title, '']
+      if (cmd.options.skip) skipToMicrobreak(false)
```
`skipToBreak` and `skipToMicrobreak` don't have any params

papey replied:
https://github.com/papey/stretchly/commit/470ba06faafad669674b590b48f08e5ae0e504d8

hovancik commented on app/main.js:
```diff
@@ -915,6 +923,47 @@ function showNotification (text) {
+  if (!cmd.checkInMain()) return
+
+  switch (cmd.command) {
+    case 'reset':
```
Can we `log.info` for each option? So in case I am debugging something I know  where events come from

papey replied:
https://github.com/papey/stretchly/commit/f7df5cb82d9df19aed4e12d846aa48d2fae0285c

hovancik commented on app/main.js:
```diff
@@ -915,6 +923,47 @@ function showNotification (text) {
+      break
+
+    default:
+      log.Error(`Unsupported command ${cmd.command}`)
```
`Error` should be lower-case, I guess.

papey replied:
https://github.com/papey/stretchly/commit/1e7b00595434ce69819118b9063385bdd1bea32a

hovancik commented:
So I believe I am pretty happy with the code and I understand what it does :) But as always with me, some things: 

1. Do you feel like writing tests? I imagine `Command` could be tested :) 
2. Have you tried with packaged version of app? I wonder whether `electron-log` outputs are visible, then. Asking because I wonder whether we should use `electron-log` or node's `log` in Command. On one hand, we would have it in logs, on the other, it's a bit uglier output (maybe that's controllable?). Right now we have a combination + Throw. 
3. Could we get rid of throw? One usage in Parse is not an issue, I think, as we are parsing from second instance, so if we would log, user will see an error in output. The other usage in `durationToMs`is harder, as it seems to be called in main instance. Could we check the value maybe when parsing? What do you think? 
4. The last think is readme, I guess only mentioning one example + saying no command is `help` would be enough? 

Let me know what you think.

papey commented:
1. Now that we have a somewhat solid base that we will not trash away, for sure I will add test to it !

2. I will test that ASAP, for me, we have to `electron-log` in the "electron app" ie, the main instance and use node's console log in the second instance.

3. In both case, the command is parse in both the main instance and the second instance, in both cases a logging line could be enough (?), I think.

4. For the README. I think a simple mention saying that you can interact with an already started and use `stretchly help` to get a list of commands should be enough.

hovancik commented:
I agree on all points :) :+1:

papey commented:
I've updated my comment to point to changes we discuss. The only thing left is testing in the packaged app. Will do it ASAP.

hovancik commented on app/main.js:
```diff
@@ -43,7 +45,10 @@ global.shared = {
 
 if (!gotTheLock) {
-  console.log('Stretchly is already running.')
+  console.log('Stretchly command instance: started\n')
```
Only wondering: why do you use `\n`? :)

papey replied:
Because I find not quite redeable when help is printed : 

```
Stretchly command instance: started
Usage: stretchly <command> [options]

Commands:
	stretchly help            Show this help message
	stretchly version         Show current stretchly version
	stretchly reset           Reset breaks
	stretchly pause [options] Pause breaks
	stretchly resume          Resume from a pause
	stretchly toggle          Toggle breaks between resume/paused
	stretchly mini [options]  Skips to next Mini Break
	stretchly long [options]  Skips to next Long Break

Options:
	-T, --title    Specify title for next break (Long or Mini)
	-t, --text     Specify text for next break (Long Break only)
	-s, --skip     Skip directly to this break (Long or Mini)
	-d, --duration Specify duration for pausing breaks (Pause only) [infinitely|until-morning|seconds (as number)]

Examples:
	stretchly pause                                    Pause breaks indefinitly
	stretchly pause -d 3600                            Pause breaks for one hour
	stretchly mini -T "Stretch up !" --skip            Skips to next Mini Break with "Stretch up!" title
	stretchly long -T "Stretch up !" -t "Go stretch !" Set next long break title to "Stretch up !" and text to "Go stretch !"
```

vs

```
Stretchly command instance: started
Usage: stretchly <command> [options]

Commands:
	stretchly help            Show this help message
	stretchly version         Show current stretchly version
	stretchly reset           Reset breaks
	stretchly pause [options] Pause breaks
	stretchly resume          Resume from a pause
	stretchly toggle          Toggle breaks between resume/paused
	stretchly mini [options]  Skips to next Mini Break
	stretchly long [options]  Skips to next Long Break

Options:
	-T, --title    Specify title for next break (Long or Mini)
	-t, --text     Specify text for next break (Long Break only)
	-s, --skip     Skip directly to this break (Long or Mini)
	-d, --duration Specify duration for pausing breaks (Pause only) [infinitely|until-morning|seconds (as number)]

Examples:
	stretchly pause                                    Pause breaks indefinitly
	stretchly pause -d 3600                            Pause breaks for one hour
	stretchly mini -T "Stretch up !" --skip            Skips to next Mini Break with "Stretch up!" title
	stretchly long -T "Stretch up !" -t "Go stretch !" Set next long break title to "Stretch up !" and text to "Go stretch !"
```

But it's a matter of taste

hovancik replied:
Ah, ok, you're right :)

papey commented:
It everything looks good to you I will finish the last detail, re-test all the stuff for a day or two, update CHANGELOG and squash everything !

hovancik commented on README.md:
```diff
@@ -92,6 +92,12 @@ Clicking the *Stretchly* icon in your tray area will display the current status
 
 <img src="dark.png" height="340">
 
+## Interact with stretchly from command line
```
can we make this h3? `###`

hovancik commented:
Looks good to me! 

I'll try dev version as well, but I trust that if you use it for your purposes it will be well tested. Let me know if packaged version is OK.

papey commented:
@hovancik just to be shure, is `npm run pack` and then running stretchly from `./dist/linux-unpacked/stretchly` enougth for testing the packaged version ? Sorry if there is some documentation about it, I did not find it.

hovancik commented:
Good question, not sure. What Linux do you run? I would edit `package.json` to only build your distribution, then run `npm run dist` and install as you would with normal installer.

papey commented:
Hey @hovancik, after a week of usage, I've rebase the branch against master, fixed a bug / weird behavior with the `long` command (see https://github.com/hovancik/stretchly/pull/700/commits/32aabbc8e769e161fedccd8bde283b24c89bbe5f). I also changed the default behavior of `long` and `mini` command. I prefer them to skip by default so `skip` option becomes `noskip`. Packaged version works for me. We should be close to squash and merge.

hovancik commented:
@papey looks good to me :) only think left: 
```
-s, --noskip   Do not skip directly to this break (Long or Mini)

```
should the short one be `-n`, or something else?

papey commented:
Was note sure about this so I left this but yeah I think you're right

papey commented:
All squashed ! :+1:

hovancik commented:
Thanks! A lot of work went into this :)

papey commented:
Thanks a lot for the feedback, was a pleasure :+1:

decadent commented:
Sorry for barging in belatedly, but it just popped up in my mind that I should take a look at the pausing command (my main interest here), and: perhaps the duration should specify minutes if it's a number? I'm very much sure that people don't usually think of work/leisure intervals in seconds. I actually have a timer app with a CLI interface (for Alfred), and I call it with the argument in minutes, which works splendidly. The list of my pause intervals in Stretchly even has a ‘30 minutes’ option, and that sounds way more natural than ‘1800 seconds’.

Personally I would also parse values in the formats ‘2h’ and ‘2h30m’, for they are also very convenient to specify on the command line. Parsing them is straightforward with regular expressions, it's basically two lines of code—I can write it if you wish. Since the commands themselves are in English, I guess the format also doesn't need internationalization.

decadent commented on app/utils/commands.js:
```diff
@@ -0,0 +1,249 @@
+  duration: {
+    long: '--duration',
+    short: '-d',
+    description: 'Specify duration for pausing breaks (Pause only) [infinitely|until-morning|seconds (as number)]',
```
‘infinitely’ here should probably be ’indefinitely’, since that's what used later in the code and in the app's menu.

papey replied:
Yeah sure, just a typo.

decadent commented on app/utils/commands.js:
```diff
@@ -0,0 +1,249 @@
+  description: 'Pause breaks for one hour'
+},
+{
+  cmd: 'stretchly mini -T "Stretch up !"',
```
A little nitpick: perhaps the command examples shouldn't have spaces before exclamation marks—like there's no spaces in the descriptions.

papey replied:
Sorry, french habits.

decadent commented on app/utils/commands.js:
```diff
@@ -0,0 +1,249 @@
+
+const allExamples = [{
+  cmd: 'stretchly pause',
+  description: 'Pause breaks indefinitly'
```
A typo: it should be ‘indefinitely’.

papey replied:
Nice catch

papey commented:
Sure it's doable. @hovancik if you want I can make a quick PR to integrate features requested by @decadent before next release

papey commented:
A starting block is available here https://github.com/papey/stretchly/tree/features/improve-pause-cmd.

hovancik commented:
Sounds good :)

hovancik commented:
@papey I'm also getting this, when Long Break is about to start: 

![image](https://user-images.githubusercontent.com/4745859/97187176-7d581f00-17a2-11eb-9c6c-96ff626b751a.png)

papey commented:
Thanks for feedback will add a quick fix

papey commented:
Since this is closed, discussion is moved to https://github.com/hovancik/stretchly/pull/748
